[2단계 - 경로 조회 기능] 로건(정다빈) 미션 제출합니다.

70825 opened the pull request:
안녕하세요 조앤 😀

이번 미션은 도메인 로직이 어렵긴하지만, 무엇보다 테스트 코드 작성할 게 정말 많았던 것 같습니다.
3단계를 슬쩍 봤는데, 금방 적용할 수 있을 것 같아요.
2단계 리뷰를 받은 후에 빠르게 3단계도 적용해보겠습니다.

잘부탁드립니다! 🙇‍♂️

---

### 1. 존재하지 않는 노선, 역을 삭제하는 상황

존재하지 않는 노선, 역을 삭제할 경우, DELETE 쿼리를 사용해도 문제가 없다는 것을 알았습니다.
그래서 일단 DELETE 쿼리를 실행한 이후에 예외륿 반환하도록 만들었는데요. 이렇게 해도 문제가 없을지 궁금합니다.

```java
public void deleteById(final Long id) {
    validateHasStaionInLine(id);

    final int count = stationDao.deleteById(id);

    if (count == ZERO) {
        throw new IllegalArgumentException("등록되지 않은 역입니다.");
    }
}
```

저는 내부 로직은 이렇게 해도 문제가 없다고 생각하는데, 코드가 예쁘게 나오지 않아서 제가 잘못된 곳에 로직을 추가한게 아닌가라는 생각이 들었습니다.

---

### 2. 최단 경로를 구하는 클래스 위치

`SectionRepository`에서 `PathService`에 의존하여 import문에서 역류가 일어나고 있습니다.
처음에는 `PathUtils`라는 클래스를 만들려고 했는데, 이 경우 Utils의 public 메서드가 전부 static으로 사용되기 때문에 객체지향 프로그래밍에서는 좋지 않아보였습니다.

```java
@Repository
public class SectionRepository {

    private final SectionDao sectionDao;
    private final StationDao stationDao;
    private final PathService pathService;

    public SectionRepository(final SectionDao sectionDao, final StationDao stationDao, final PathService pathService) {
        this.sectionDao = sectionDao;
        this.stationDao = stationDao;
        this.pathService = pathService;
    }

   ....
}
```

이 부분은 `PathService`를 그대로 둔 상태에서 `SectionRepository`를 사용하는 `LineCommandService`에서 미리 경로를 구하는게 더 좋을지, 아니면 더 좋은 방법이 있는지 궁금합니다.

---

### 3. 통합 테스트와 인수 테스트의 차이

제가 만든 통합 테스트 패키지에 있는 테스트 클래스가 통합 테스트인지, 인수 테스트인지 잘 모르겠습니다.
둘의 차이를 검색 해봐도 둘이 목적만 다르다 나와있고, 우테코 4기 분들의 지하철 미션 인수 테스트를 보니 통합 테스트와 흡사해서 어느 점이 코드상에서 다른 것인지 잘 모르겠습니다. 
[Tecoble 글](https://tecoble.techcourse.co.kr/post/2021-05-25-unit-test-vs-integration-test-vs-acceptance-test/)을 읽어보면 그러면 E2E 테스트랑 인수 테스트랑은 무슨 차이가 있는 것인지 모르겠더라구요.

제가 이해하기로 통합 테스트와 인수 테스트는 개발자 입장에서는 비슷한 코드지만, 직접 간단한 시나리오를 만들어서 테스트하거나(통합 테스트), 다른 회사 직원분한테 시나리오를 받아서 테스트하거나(인수 테스트)의 차이인 것인가요?
혹시 조앤은 차이점을 아시고 계신지 궁금합니다.

---

감사합니다!

70825 committed:
refactor: 1단계 리팩터링

70825 committed:
docs: 2단계 기능 목록 추가

70825 committed:
build: jgrapht 추가

70825 committed:
feat: 노선의 경로를 정렬하는 기능 추가

70825 committed:
build: jgrapht에서 jgrapht-core로 변경

70825 committed:
feat: 경로를 조회하는 기능 추가

70825 committed:
feat: 요금 계산 기능 추가

70825 committed:
style: 코드 포맷 수정

70825 committed:
build: mysql-connector-java 추가

70825 committed:
feat: 데이터베이스 설정 분리

70825 committed:
refactor: 경로와 요금 API 통합

70825 committed:
refactor: 메서드 순서 변경

70825 committed:
fix: 경로 조회 API에 불필요한 정보 제거, 테스트 코드 수정

70825 committed:
build: validation 추가

70825 committed:
feat: request dto에 validation 적용

70825 committed:
refactor: 예외 핸들러, 요청 DTO 네이밍 수정

70825 committed:
fix: 중복된 구간을 등록할 경우, 제대로된 예외를 반환하도록 수정

70825 committed:
test: 부족한 테스트 추가

70825 committed:
feat: 존재하지 않는 노선, 역을 삭제하려고 하면 예외 메시지를 반환하도록 추가

70825 committed:
refactor: API 반환값 이름 변경

70825 committed:
style: 코드 포맷 수정

70825 committed:
build: Swagger 추가

70825 committed:
feat: API 문서 기능 추가

70825 committed:
refactor: 오탈자 수정

seovalue commented on README.md:
```diff
@@ -73,7 +80,10 @@ CREATE TABLE IF NOT EXISTS section
-3. API 문서
+### 3. API 문서
+
+[Swagger 문서](http://localhost:8080/swagger-ui/index.html)
```
👍 swagger를 적용해주시니 확인하기가 편리하네요~ RestDocs도 있으니 나중에 활용해보세요ㅎㅎ 테스트 기반으로 동작해요

70825 replied:
이번에 페어와 같이 시간이 남으면 RestDocs를 적용하기로 했었는데, 서로 도메인을 수정하느라  바빠서 못했던 것 같아요 ㅠㅠ
시간이 된다면 마지막 제출 전에 변경해보겠습니다!

seovalue commented on src/main/java/subway/ui/dto/request/CreationSectionRequest.java:
```diff
@@ -0,0 +1,42 @@
+
+public class CreationSectionRequest {
+
+    @Schema(description = "상행 종점 방향 역 ID")
+    @NotNull
+    private final Long upStationId;
+
+    @Schema(description = "하행 종점 방향 역 ID")
+    @NotNull
+    private final Long downStationId;
```
상행과 하행이 같은 요청을 보냈을 때 
```
2023-05-19 22:40:32.037 ERROR 28191 --- [io-8080-exec-10] subway.ui.GlobalExceptionHandler         : Class : IllegalArgumentException, Message : loops not allowed
```
라는 응답이 오는데 케이스를 잡아서 적절한 예외를 던져보면 어떨까요? 해당 메시지는 조회는 안되는데 어디서 발생하는 건지.. 👀

70825 replied:
저 부분은 jgrapht에서 엣지에서 서로 다른 두 개의 정점이 들어와야 하는데, 같은 정점이 들어올 때 발생하더라구요.
수정했습니다!

seovalue commented on src/main/java/subway/ui/StationController.java:
```diff
@@ -26,18 +30,23 @@ public StationController(final StationService stationService) {
-    public ResponseEntity<CreationStationResponse> createStation(@RequestBody final CreationStationRequest request) {
-        final CreationStationDto stationDto = stationService.saveStation(request.getName());
-        final CreationStationResponse response = CreationStationResponse.from(stationDto);
+    @Operation(summary = "역 생성 API", description = "새로운 역을 생성합니다.")
```
swagger를 사용하게 되면 코드에 문서와 관련된 annotation들이 침투한다는 단점이 있어요. 이 부분에 대해서는 어떻게 생각하세요?

70825 replied:
만약 API 문서를 컨트롤러에 꼼꼼하게 적는다면 코드 읽기가 굉장히 힘들 것 같아요.
인터페이스에 swagger 관련 내용을 추가하는 방법도 있다하던데, 차라리 초반에 고생 좀 해서 rest docs를 만드는게 더 좋아보여요

70825 replied:
다음 미션에 rest docs를 적용해보니까 API 문서의 신뢰도를 생각한다면 swagger를 사용하며 꼼꼼하게 관리하는 것보다 restdocs를 사용하는게 좋을 것 같아요
swagger의 장점은 rest docs에는 없는 요청과 응답을 바로 테스트 해볼 수 있다는 점인데, 검색해보니 swagger + restdocs를 사용하는 방법도 있더라구요. 그래서 swagger가 빠르고 쉽게 API 문서를 만드는게 좋다는 장점도 있지만, 웬만하면 rest docs를 적용하는게 좋은 것 같습니다!

seovalue commented on src/main/java/subway/application/LineCommandService.java:
```diff
@@ -0,0 +1,64 @@
+        return lineRepository.insert(line);
+    }
+
+    public Line findLineById(final Long id) {
+        final Line line = lineRepository.findById(id);
+        return sectionRepository.findAllSectionByLine(line);
+    }
```
Command와 Query로 서비스를 구분해주신 기준이 궁금해요~

70825 replied:
장바구니 미션을 진행할 때, 다른 크루들이 CQRS 패턴에 대해 많이 언급을 하기도 하고, 저번 미션때 테스트 수행 시간을 최적화하면서 참고했던 [테코톡 영상](https://youtu.be/N06UeRrWFdY?t=560)에서도 이런 내용이 있길래 간단하게 학습을 했었어요
때마침 지금 미션이 사용자들이 사용하게 된다면 CUD는 잘 일어나지 않고, 조회하는 일이 대다수라 이 방법을 사용해도 좋을 것 같아서 적용했습니다! 
디비도 분리해서 적용해보고 싶었는데,  이 부분은 따로 학습이 필요해보여서 적용을 못했어요

seovalue commented on src/main/java/subway/application/LineCommandService.java:
```diff
@@ -0,0 +1,64 @@
+        final Station downStation = stationRepository.findById(downStationId);
+        final Section section = Section.of(upStation, downStation, Distance.from(distance));
+
+        final Line newLine = sectionRepository.findAllSectionByLine(line);
```
여기도 그렇고 findAllSectionByLine 내부도 그렇고 동일한 Line을 한번 더 재생성해서 newLine이라고 칭한 뒤, 해당 newLine에 섹션을 추가하는 방식으로 동작하는데 이유가 있을까요?

70825 replied:
생각해보니 이렇게 해야할 이유가 없는 것 같아요
그래서 lineId를 통해 노선에 저장된 구역까지 한 번에 가져오도록 수정했습니다.

seovalue commented on src/main/java/subway/domain/section/delete/strategy/DeleteStationStrategy.java:
```diff
@@ -0,0 +1,10 @@
+import subway.domain.section.Sections;
+
+@FunctionalInterface
+public interface DeleteStationStrategy {
```
삭제에 대해 전략패턴을 사용해주셨군요?? 도입하시게 된 계기가 궁금해요!!

70825 replied:
처음 계기는 1단계를 진행할 때부터 삽입/삭제 방법이 최소 2가지 이상이라 전략 패턴을 사용해도 적절해보였어요
그래서 2단계 미션을 진행하기 전에 전략 패턴을 적용하여 리팩터링을 진행했는데, 막상 적용해보니까 장단점이 있는 것 같아요

[장점]
한 도메인 클래스에 있는 코드가 너무 비대했었는데, 전략패턴 덕분에 코드 수가 많이 줄어들었어요. 전략 패턴을 적용한 뒤에 메서드를 삭제하는 과정에서 최소 100라인은 넘는 코드들이 삭제된 것으로 기억해요

[단점]
도메인에 넣어두었다면 접근 제어자를 private로 해도 충분했을법한 메서드들이 전부 public으로 되었다는 점이 단점 같아요

저는 private로 해도 될 메서드를 전부 public으로 변경 했던게 큰 단점으로 보여져서 고치고 싶지만, 이걸 해결하기 위해 삽입/삭제 세부 로직을 도메인에 전부 넣어버리면 다시 도메인 코드가 너무 비대해져서 어떻게 해야할지 판단이 안서네요..
조앤이 이런 상황이라면 어떻게 하시는 편인가요?

seovalue replied:
저는 우선 구현해주신 것을 보면 장점이 좀 더 와닿는 것 같아요! 말씀해주신 것처럼 전략을 사용하지 않으면 이 모든 코드가 한 서비스나 특정 클래스에 모두 들어가게 되어서 비즈니스 로직이 굉장히 길어지는데, 분리되어서 확인할 수 있어서 가독성이 더 좋은 것 같아요. 저는 주로 코드를 짤 때 가독성을 가장 중요하게 생각해요!

seovalue commented on src/main/java/subway/application/PathService.java:
```diff
@@ -0,0 +1,23 @@
+
+import java.util.List;
+
+public interface PathService {
```
왜 이친구만 interface를 사용해주셨나요?

70825 replied:
여기는 JGraphT 라이브러리에 의존하기 때문에 다른 클래스들보다 교체될 가능성이 높아보여서 사용했습니다!
그런데 이런건 규칙성 있게 하나를 적용하면 나머지도 적용하는게 맞겠죠..??

seovalue replied:
처음부터 interface를 도입하는 것과 추후 interface를 빼내어 분리하는 것은 장단점이 존재하는 것 같아요. 적절히 사용해주시면 좋을 것 같아요~~ㅎㅎ 구현체가 1개인데 인터페이스를 사용해주신 이유가 궁금해서 여쭤봤어요~

seovalue commented on src/main/java/subway/application/PathServiceImpl.java:
```diff
@@ -0,0 +1,76 @@
+    public List<Section> getSectionsByShortestPath(final Station sourceStation,
+                                                   final Station targetStation,
+                                                   final List<Line> lines) {
+        return getShortestPath(sourceStation, targetStation, lines).getEdgeList().stream()
+                .map(edge -> Section.of(edge.getSource(), edge.getTarget(), Distance.from(edge.getDistance())))
+                .collect(Collectors.toList());
```
따지고보면 한줄인 return문에 `.`이 몇개일까요...ㅎㅎ 체이닝이 여러번 되면 가독성을 떨어트리는 것 같아요!

70825 replied:
코멘트 주신 부분이랑 다른 부분도  .이 너무 많은 부분이 있어서 같이 수정했습니다!

seovalue commented on src/main/java/subway/application/PriceService.java:
```diff
@@ -0,0 +1,60 @@
+    private static final Distance EIGHT_KM_DISTANCE = Distance.from(8);
+    private static final Distance TEN_KM_DISTANCE = Distance.from(10);
+    private static final Distance FIFTY_KM_DISTANCE = Distance.from(50);
+    private static final Distance MAXIMUM_DISTANCE = Distance.from(1_000_000);
```
만약 10\~12km, 12\~14km 이렇게 구간이 세부적으로 나뉜다면 계속 늘어날거예요ㅎㅎ
패턴을 잘 적용해주시는 것 같아 [상태패턴](https://tecoble.techcourse.co.kr/post/2021-04-26-state-pattern/)과 관련된 내용도 링크해요! 한번 적용해보시면 어떨까요?
상태패턴이 아니더라도, 변경이 잦다고 생각하고 유연하게 리팩터링해보시는게 어떨까요~

70825 replied:
구간 수정, 삭제를 했던 전략 패턴이랑 비슷하게 만들어서 적용했습니다!
전략 패턴이랑 다른 점은 가격을 누적해서 더할 수 있도록 만들었습니다.

seovalue commented on src/main/java/subway/domain/section/add/strategy/MidAddSectionStrategy.java:
```diff
@@ -0,0 +1,18 @@
+
+    @Override
+    public void addSection(final Sections sections, final Section insertSection) {
+        final Section originSection = sections.findOriginSection(insertSection);
+        final Section upSection = insertSection.createUpSection(originSection);
+        final Section downSection = insertSection.createDownSection(originSection);
+        final int originIndex = sections.findIndex(originSection);
+        sections.delete(originSection);
+        sections.addSection(originIndex, upSection);
+        sections.addSection(originIndex + 1, downSection);
```
이부분도 유사한 역할을 하는 부분끼리 묶어 메서드로 한번 더 분리해봐도 좋을 것 같아요! 흐름이 읽히도록요

70825 replied:
이 부분은 메서드로 분리하면 파라미터가 너무 많이 생겨서 어떻게 분리해야할지 모르겠네요..
`Section`의 `calculateDistance` 메서드처럼 `createUpSection`이랑 `createDownSection`을 `createSection`으로 합치는 것도 생각해봤는데, 그러면 createSection에 파라미터가 많이 생겨서 줄바꿈으로 구분했습니다!

```suggestion
        final Section originSection = sections.findOriginSection(insertSection);
        final int originIndex = sections.findIndex(originSection);
        sections.delete(originSection);

        final Section upSection = insertSection.createUpSection(originSection);
        sections.addSection(originIndex, upSection);

        final Section downSection = insertSection.createDownSection(originSection);
        sections.addSection(originIndex + 1, downSection);
```

seovalue commented on src/main/java/subway/domain/section/Section.java:
```diff
@@ -0,0 +1,130 @@
+            throw new IllegalArgumentException("현재 구역의 역이 삭제되지 않았습니다. 삭제하려는 구역의 역이 현재 구역의 역에 존재하는지 확인해주세요");
+        }
+
+        return ownStations.iterator().next();
```
단 2개임이 당연해서 next()로 구간의 나머지 역을 반환하는 것 같은데요, 명시적으로 나타낼 수 있으면 좋을 것 같아요.

70825 replied:
otherStation 같은 변수를 따로 생성하고, otherStation을 반환하는 것으로 이해한게 맞을까요..??
일단 이렇게 수정했습니다!

seovalue commented on src/test/java/subway/persistence/dao/SectionDaoTest.java:
```diff
@@ -34,18 +29,38 @@ class SectionDaoTest extends DaoTest {
         final List<SectionEntity> actual = sectionDao.findAll();
 
+        // then
+        SoftAssertions.assertSoftly(softAssertions -> {
```
SoftAssertions를 사용해주신 이유가 무엇인가요?

70825 replied:
assertAll보다 어디 부분에서 에러가 나왔는지 좀 더 파악하기 쉬운 것 같아서 적용했습니다!

### assertAll

<img width="382" alt="assertable" src="https://github.com/woowacourse/jwp-subway-path/assets/79046106/5705cd15-8979-4c02-93cb-e109d4c7e1df">

### assertSoftly

<img width="488" alt="softassertly" src="https://github.com/woowacourse/jwp-subway-path/assets/79046106/6e5db07d-144a-4622-89b6-e6eb0cba90ff">

seovalue commented on src/test/java/subway/integration/PathIntegrationTest.java:
```diff
@@ -0,0 +1,213 @@
+        final CreationStationRequest stationRequestFive = CreationStationRequest.from("미금역");
+        final CreationStationRequest stationRequestSix = CreationStationRequest.from("선릉역");
+
+        // 노선 등록
+        final ExtractableResponse<Response> createLineResponseOne = postLine(lineRequestOne);
+        final ExtractableResponse<Response> createLineResponseTwo = postLine(lineRequestTwo);
+
+        // 지하철역 등록
+        final ExtractableResponse<Response> createStationResponseOne = postStation(stationRequestOne);
+        final ExtractableResponse<Response> createStationResponseTwo = postStation(stationRequestTwo);
+        final ExtractableResponse<Response> createStationResponseThree = postStation(stationRequestThree);
+        final ExtractableResponse<Response> createStationResponseFour = postStation(stationRequestFour);
+        final ExtractableResponse<Response> createStationResponseFive = postStation(stationRequestFive);
+        final ExtractableResponse<Response> createStationResponseSix = postStation(stationRequestSix);
+
+        // id값 확인
+        lineOneId = getParseId(createLineResponseOne);
+        lineTwoId = getParseId(createLineResponseTwo);
+        stationOneId = getParseId(createStationResponseOne);
+        stationTwoId = getParseId(createStationResponseTwo);
+        stationThreeId = getParseId(createStationResponseThree);
+        stationFourId = getParseId(createStationResponseFour);
+        stationFiveId = getParseId(createStationResponseFive);
+        stationSixId = getParseId(createStationResponseSix);
+
+        // 구간 요청 데이터 세팅
+        final CreationSectionRequest sectionRequestOneLineOne = CreationSectionRequest.of(stationOneId, stationTwoId, 10);
+        final CreationSectionRequest sectionRequestOneLineTwo = CreationSectionRequest.of(stationTwoId, stationThreeId, 20);
+        final CreationSectionRequest sectionRequestOneLineThree = CreationSectionRequest.of(stationThreeId, stationFourId, 20);
+        final CreationSectionRequest sectionRequestOneLineFour = CreationSectionRequest.of(stationFourId, stationFiveId, 40);
+        final CreationSectionRequest sectionRequestTwoLineOne = CreationSectionRequest.of(stationFourId, stationSixId, 8);
+        final CreationSectionRequest sectionRequestTwoLineTwo = CreationSectionRequest.of(stationSixId, stationFiveId, 8);
+
+        // 구간 등록
+        postSection(lineOneId, sectionRequestOneLineOne);
+        postSection(lineOneId, sectionRequestOneLineTwo);
+        postSection(lineOneId, sectionRequestOneLineThree);
+        postSection(lineOneId, sectionRequestOneLineFour);
+        postSection(lineTwoId, sectionRequestTwoLineOne);
+        postSection(lineTwoId, sectionRequestTwoLineTwo);
```
와우 이런것 다른 클래스에서도 공통으로 사용될 수 있는 변수들이 보이는데 Fixture를 만들어두고 사용해봐도 좋을 것 같아요.
 예를 들면 LineFixture 안에는 Line과 관련된 요청들이 들어있다던지요ㅎㅎ

70825 replied:
Fixture라는 패키지를 만들었습니다. 그런데 한 곳에 넣기엔 유지보수하기엔 안좋아보여서 각 도메인, DTO마다 하나의 파일에서 관리하도록 변경했습니다!

seovalue commented on src/test/java/subway/integration/PathIntegrationTest.java:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    @Test
+    void 구간_10km_50km의_과금과_50km이상의_과금을_포함한_경로와_요금을_가져오는지_확인한다() {
```
요런건 도메인 테스트를 통해 커버해보는 것도 좋은 것 같아요.

seovalue commented on src/test/java/subway/ui/dto/request/CreationLineRequestTest.java:
```diff
@@ -0,0 +1,46 @@
+package subway.ui.dto.request;
```
dto 테스트를 해주신 이유가 궁금합니다!

70825 replied:
다른 dto는 getter만 사용하기 때문에 굳이 테스트를 할 필요가 없지만, request dto는 NotNull, NotBlank인지 확인하는 중요한 역할을 하기 때문에 추가했습니다!

seovalue commented on src/main/java/subway/persistence/repository/SectionRepository.java:
```diff
@@ -1,118 +1,60 @@
-            }
+        final Station upStation = line.getSections().findUpSection().getUpStation();
+        final Station downStation = line.getSections().findDownSection().getDownStation();
+        final List<Section> orderedSectionPath = pathService.getSectionsByShortestPath(upStation, downStation, List.of(line));
```
우선 insert를 위해 불필요한 삭제가 이뤄진다는 점이 어색하게 느껴져요!


2. 최단경로를 구하는 클래스 위치
제가 이해하기론 결과적으로 line의 구간 순서대로 DB에도 삽입하기 위해 pathService가 레포지토리까지 침투한게 아닌가 싶은데요ㅎㅎ(혹시 아니라면 알려주세요ㅎㅎ 머쓱) 굳이 순서대로 DB에 저장하지 않고 조회 시 정렬해도 될 것 같구요, DB에 꼭 순서대로 저장해야하는 이유가 있다면 서비스에서 최종 정렬까지 된 값이 repository에 넘어와서 repository는 단순 insert 역할만 도와야하지 않을까 싶어요~

70825 replied:
line이 가지고 있는 section을 삭제하고, 다시 순서대로 저장하는 이유는 역과 역 사이에 새로운 역을 추가하거나, 중간에 있는 역을 삭제할 때 Repository의 로직이 복잡하다고 느껴졌습니다.

저도 조앤의 의견이 맞다고 생각해서 조회할 때 정렬하는 기능을 구현하려고 했는데, 순서대로 구역이 들어오지 않으면 어느 역이 상행종점역인지, 하행종점역인지 알 수 없더라구요 😢
JGraphT로 알 수 있을까 싶어서 검색도 해보고, 내부 메서드들도 확인했는데 제가 찾아보기로는 끝과 끝 사이 노드를 구하는 메서드가 없었습니다.

그래서 트레이드 오프처럼 생각해서 어쩔 수 없이 서비스에서 정렬된 값이 repository로 넘어오도록 수정했습니다!

seovalue commented on README.md:
```diff
@@ -2,42 +2,49 @@
-- [ ] 노선도 (Route Map)
-  - [ ] 노선도는 노선들을 가지고 있다
-    - [ ] 노선의 이름과 색깔은 각각 고유한 값이어야 한다
+#### 1단계 기능 목록
```
### 존재하지 않는 노선, 역을 삭제하는 상황

저는 지금 구현해주신 대로 해도 문제가 없고, 없는 것을 삭제하는 요청이 들어왔을 때 그냥 별다른 예외를 던져주지 않아도 상관없다고 생각해요.

없는 걸 삭제해서 없다고 예외를 던졌을 때 그 예외를 받고 추가적으로 할 수 있는 액션이 무엇이 있을지? 고민해보시면 좋을 것 같아요.
아니면 있는지 조회를 먼저 한 뒤 없으면 예외를 반환해도 될 것 같구요.

### 최단 경로를 구하는 클래스 위치

는 해당 코드에 코멘트 남길게요~~

### 통합 테스트와 인수 테스트의 차이

저는 솔직히 어떻게 말하냐의 차이라고 생각하긴 해요ㅎㅎ 실제 api를 호출해서 DB까지 단 하나도 mocking하지 않은 실 동작을 테스트 해보는 것을 통합테스트이자 인수테스트라고 생각합니다! 

인수테스트라는 용어가 등장한게 사실상 비개발자분들도 이 테스트를 통해 비즈니스 요구사항이 제대로 동작하는 지 확인할 수 있는 명세같은 역할을 할 수 있도록 하는 테스트라는 의미를 좀 더 담기 위해 이런 용어를 사용하는 거지 않을까 싶어요.. (실제로 비개발자분들은 전혀 코드를 보지 않지만요..ㅋㅋ) 예를 들면 cucumber라는 것을 활용하면 아래처럼 테스트를 짤 수 있어요.
```
시나리오: Breaker guesses a word
  조건 the Maker has chosen a word
  만약 the Breaker makes a guess
  그러면 the Maker is asked to score
```

이렇게 짠다면 이걸 돌려보고 명세대로 동작하고 이게 정말 우리 서비스의 요구사항에 대한 명세이자 테스트가 될 수 있다고 생각할 수 있을 것 같아요. 하지만 이것도 표현의 차이이지 내부적으로는 다 구현되어야하거든요. 그냥 말의 차이인 것 같아요 저는?!!?

seovalue commented on src/main/java/subway/application/PriceService.java:
```diff
@@ -0,0 +1,60 @@
+        return BASIC_FARE.plus(fareForTenToFiftyKm).plus(fareForOverFiftyKm);
+    }
+
+    private Price calculateAdditionFare(final Distance distance,
+                                        final Distance minimumDistance,
+                                        final Distance maximumDistance,
+                                        final Distance distancePerFare) {
+        if (distance.isLessThanOrEqualTo(minimumDistance)) {
+            return NO_ADDITION_FARE;
+        }
+
+        final Distance remainDistance = getDistanceForIntervalCalculating(distance, minimumDistance, maximumDistance);
+
+        return Price.from(calculateSurcharge(remainDistance, distancePerFare));
+    }
+
+    private Distance getDistanceForIntervalCalculating(final Distance distance,
+                                                       final Distance minimumDistance,
+                                                       final Distance maximumDistance) {
+        final int originValue = distance.getDistance();
+        final int minimumValue = minimumDistance.getDistance();
+        final int maximumValue = maximumDistance.getDistance();
+
+        if (originValue > maximumValue) {
+            return Distance.from(maximumValue - minimumValue);
+        }
+
+        return Distance.from(originValue - minimumValue);
+    }
+
+    private int calculateSurcharge(final Distance distance, final Distance farePerDistance) {
+        final int originValue = distance.getDistance();
+        final int perFareValue = farePerDistance.getDistance();
+        final int additionFare = ADDITION_FARE.getPrice();
+
+        return (((originValue - 1) / perFareValue) + 1) * additionFare;
+    }
```
추가로 이런게 도메인안으로 숨겨진다면 도메인 레이어에서 요금정책과 관련된 것을 모두 테스트할 수 있을 것 같아요~

70825 replied:
요금 정책마다 `getDistanceForIntervalCalculating`를 사용하는 곳이 하나뿐이라 `calculateSurcharge`만 Price로 이동하고, 나머지는 calculator 패키지 안에서 계산할 수 있도록 변경했습니다!

seovalue reviewed (changes requested):
안녕하세요, 로건! 조앤이에요ㅎㅎ
몸은 좀 괜찮아지셨나요. 2단계 잘 수행해주셨네요ㅎㅎ 
리뷰 남겼는데 남기다보니 좀 많아졌어요😅 질문에 대한 답변도 남겼어요. 
확인해보시고 다시 리뷰요청주세요~

70825 committed:
refactor: jsoncreator 삭제

70825 committed:
fix: 상행과 하행이 같을 때, 다른 예외 메시지를 반환하도록 수정

70825 committed:
refactor: 체이닝 줄이기

70825 committed:
refactor: 공통된 변수는 상수로 변경

70825 committed:
refactor: 노선 id를 통해 구역을 가진 노선을 반환할 수 있도록 수정
AS-IS: 기존에는 LineRepository에서 노선 id를 검색 후, SectionRepository에서 구역을 추가함
TO-BE: 지금은 노선 id를 통해 구역까지 한 번에 가져오도록 수정

70825 committed:
refactor: 줄바꿈을 통해 코드 흐름을 구분하도록 변경

70825 committed:
refactor: 한 구역에서 하나의 역이 삭제될 때, 나머지 역을 명시적으로 나타내도록 변경

70825 committed:
refactor: 요금 정책을 쉽게 수정할 수 있도록 변경

70825 committed:
refactor: PathService 인터페이스 삭제

70825 committed:
test: 요금 계산 테스트 추가

70825 committed:
refactor: 순서가 있는 경로를 구하는 로직의 위치를 이동

70825 commented:
안녕하세요 조앤 리뷰 요청이 많이 늦어서 죄송합니다 😭
지하철 미션 난이도가 어려워서 그랬던 것인지, 감기로 인해 페이스가 말렸던 것인지, 번아웃이 왔었습니다
우테코를 하면서 레벨마다 힘든 시기가 있는데, 세번째 미션마다 집중력이 확 떨어지는 것 같아요
조앤에게 리뷰를 더 받았으면 좋을 것 같은데 아쉽습니다 ㅠㅠ

seovalue commented on src/main/java/subway/domain/Distance.java:
```diff
@@ -35,8 +37,25 @@ public Distance minus(final Distance other) {
 
-    public boolean isLessThanOrEqualTo(final Distance value) {
-        return distance <= value.getDistance();
+    public boolean isLessThanOrEqualTo(final Distance other) {
+        return distance <= other.distance;
+    }
+
+    public boolean isMoreThan(final Distance other) {
+        return this.distance > other.distance;
+    }
```
다른 distance 객체와 비교하는거 좋네요 👍

seovalue commented on src/main/java/subway/domain/price/PriceCalculatorHandler.java:
```diff
@@ -0,0 +1,29 @@
+
+import java.util.Arrays;
+
+public enum PriceCalculatorHandler {
```
enum으로 처리해주셨군요 👍

seovalue commented on src/test/java/subway/fixtures/domain/SectionFixture.java:
```diff
@@ -0,0 +1,40 @@
+import static subway.fixtures.domain.StationFixture.SEOLLEUNG_DOWN;
+import static subway.fixtures.domain.StationFixture.SEOLLEUNG_MID;
+
+public class SectionFixture {
```
👍

seovalue commented on src/test/java/subway/domain/price/calculator/BasicPriceCalculatorTest.java:
```diff
@@ -0,0 +1,22 @@
+
+@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
+@SuppressWarnings("NonAsciiCharacters")
+class BasicPriceCalculatorTest {
```
각 Calculator도 테스트를 해주셨군요 👍

seovalue reviewed (approved):
안녕하세요, 로건! 조앤이에요ㅎㅎ
리팩터링 잘 수행해주셨군요 👍 요구사항은 모두 만족하는 것 같아 여기서 머지해도 될 것 같아요ㅎㅎ
마지막 미션이 끝난 뒤 방학때엔 여유를 갖고 푹 쉬시길 바랄게요
고생 많으셨어요~
